### DISCUSSION TITLE
First run left me with more questions than answers

### DISCUSSION THREAD

**Discussion by astrojuanlu** [pdm-project/pdm#919] [General]
This is a recollection of my first run of `pdm`:

1. I did `pipx install pdm`, so far so good
2. Before changing anything globally, I decided (perhaps mistakenly) to skip the "enable PEP 582 globally" section of [the docs](https://pdm.fming.dev/) (more on that below).
3. I ran `pdm init` (good) and `pdm add streamlit`. This led to lots of backtracking, probably because it was using the system-wide `/usr/bin/python3.9` interpreter ❗ so I `Ctrl+C` 'ed.
4. I then created a venv to avoid this issue, `python3.9 -m venv .venv && source .venv/bin/activate`, and ran `pdm add streamlit` again. Initially I didn't set any lower bound to the Python version, which led to #917. Then I tried `--dry-run` and got #918. Eventually I got it installed though, so let's continue.
   - The first one was a misunderstanding on my side and the second one was a minor bug that was promptly fixed.
5. I fired a Python interpreter and tried `import urllib3` (which is a dependency of streamlit), but got a "module not found". I mistakenly assumed that pdm would install the dependencies in the `.venv` if instructed to do so (since step 4, `.pdm.toml` has the `.venv` interpreter as the `path`) 😕
6. I went back to the docs, because clearly I was going too fast and I thought I would need to enable PEP 582. Running `export PYTHONPATH='/src/.venv/lib/python3.9/site-packages/pdm/pep582'` made `import urllib3` work 🏁🥵 

It was a rough ride, but now I think I got it. Things that in my opinion could be improved:

- If `pdm` advertises itself as a package manager that "doesn't need to create a virtualenv at all", maybe the default behavior when using a system-wide interpreter should be exactly the same as a fresh venv? Or maybe I'm misunderstanding something?
- If used inside a `venv`, perhaps some feedback could be given to the user that PEP 582 is not enabled so they don't get surprised that packages don't get installed there?
- Alternatively, if used inside a `venv`, PEP 582 is not necessary at all, and the packages could be installed in the `venv` directly.
- In the same way that `pdm --pep582` does all the work magically on Windows, perhaps there could be a command that works only for the current shell, in case folks want to experiment before changing their `.bash_profile`? 

What do you think?

**Comment by pawamoy** [comment]
> If pdm advertises itself as a package manager that "doesn't need to create a virtualenv at all", maybe the default behavior when using a system-wide interpreter should be exactly the same as a fresh venv? Or maybe I'm misunderstanding something?

It should, yes. Using a system interpreter is a correct thing to do. You probably saw lots of backtracking because of the empty "requires-python". I just tried myself, using Python 3.8 and requires-python >= 3.8, pdm add streamlit and everything went fine.

> If used inside a venv, perhaps some feedback could be given to the user that PEP 582 is not enabled so they don't get surprised that packages don't get installed there?
> Alternatively, if used inside a venv, PEP 582 is not necessary at all, and the packages could be installed in the venv directly.

From my understanding, when a venv is activated, PDM is indeed supposed to install things in it, and therefore python should find installed packages and you should be able to launch scripts from the CLI. But I've never used PDM with venvs so I can't tell for sure.

> In the same way that pdm --pep582 does all the work magically on Windows, perhaps there could be a command that works only for the current shell, in case folks want to experiment before changing their .bash_profile?

There's `pdm run` 😄

**Reply by astrojuanlu** [reply]
> It should, yes. Using a system interpreter is a correct thing to do. You probably saw lots of backtracking because of the empty "requires-python". I just tried myself, using Python 3.8 and requires-python >= 3.8, pdm add streamlit and everything went fine.

You're absolutely right, just came to comment that same thing.

> From my understanding, when a venv is activated, PDM is indeed supposed to install things in it, and therefore python should find installed packages and you should be able to launch scripts from the CLI. But I've never used PDM with venvs so I can't tell for sure.

Should I open an issue about this then?

> There's `pdm run`

Didn't know about `pdm run`, thanks! 🚀

**Reply by astrojuanlu** [reply]
The answer was in the docs: https://pdm.fming.dev/usage/project/#working-with-a-virtualenv 🎉
